### ISSUE TITLE
Omniscient AI Point System for evaluating the AI's mons

### ISSUE THREAD

**Issue by ShadowzLmao2** [rh-hideout/pokeemerald-expansion#4879]
### Description

In competitive Pokemon, we see that pro players will evaluate their team against their opponent. They will look at their mons and see how many of the opponents mons they can one-shot and how many they can hard wall or check. This comes into play when the AI_FLAG_OMNISCIENT is used but can be used based on info gathered over the course of the battle. The idea is that the AI gives each mon a score in each category of Check (Outspeed and kill) and Counter (take very little damage from). When there are checks like seeing how much BST a mon has to determine whether or not it's useful, we can replace that with this system. If the AI knows it's mon will die but it can check an opposing mon, it will instead try to sacrifice a mon that can check/counter nothing on the player team.
Using merit based system instead of a static BST or HP check is better because a mon can have bad stats or low HP but is still able to check or wall certain mons (like for example ferroseed being useful). If you would like help with ideas for implementation lmk on discord. I dont get github notifications. DM is fine, ping on RHHideout is also fine

### Discord contact info

SShadowzLmao#3643

**Comment by AlexOn1ine** [comment]
Obviously a good idea but I worry that this will slow down the AI significantly

**Comment by Pawkkie** [comment]
I've been thinking about this for a few days.

> The idea is that the AI gives each mon a score in each category of Check (Outspeed and kill) and Counter (take very little damage from). When there are checks like seeing how much BST a mon has to determine whether or not it's useful, we can replace that with this system.

As far as I'm aware at least, there isn't a pre-existing determination for whether a mon is "useful". Revenge killing (checks) and defensive mons (counters) are both handled by `AI_FLAG_SMART_MON_CHOICES`, and the determination to switch in a counter is made by `AI_FLAG_SMART_SWITCHING`. With these flags active, the AI should switch in strong counters mid-battle where appropriate, and revenge killers post-KO where appropriate.

> Using merit based system instead of a static BST or HP check is better because a mon can have bad stats or low HP but is still able to check or wall certain mons (like for example ferroseed being useful).

There isn't an explicit system using those values as checks as far as I'm aware of, as the function that determines a mon's suitability in the current context, `GetBestMonIntegrated`, factors in a *lot* more considerations.

>  If the AI knows it's mon will die but it can check an opposing mon, it will instead try to sacrifice a mon that can check/counter nothing on the player team.

As far as I'm interpreting this suggestion, this is the primary functionality that is not implemented in any way, and would need to be to approximate this merit-based evaluation without obliterating the AI calcs. Is that a fair assessment?

**Comment by Pawkkie** [comment]
Been thinking about this as I've been going through some of the battle-ai labeled issues, with a focus on the concept of sacrifice fodder because that's the big thing missing.

The best implementation I can come up with (this is all pseudocode) is adding some struct field to track whether each of the AI's mons is considered fodder against each of the player's mons
```
battlerIsFodder[PARTY_SIZE][PARTY_SIZE]
```
and then, at battle start, doing an obscenely expensive calculation to determine whether each of the AI's party mons can win a 1v1 against each of the player's mons
```
for (mon in AI party)
{
	for (mon in player party)
	{
		for (move in AI mon)
		{
			if (CanWin1v1(aiMon, playerMon, move)
				battlerIsFodder[aiPartyIndex][playerPartyIndex] = FALSE;
			else
				battlerIsFodder[aiPartyIndex][playerPartyIndex] = TRUE;
		}	
	}
}
```
then when looking for switchin candidates via `GetMostSuitableMonToSwitchInto`, checking whether the AI mon is fodder and adding a new ID to `GetBestMonIntegrated` to track it
```
for (mon in player party)
{
	if (!battlerIsFodder[aiPartyIndex][playerPartyIndex])
		continue;
	else
		fodderId = aiPartyIndex;
}
```
and adding a corresponding `ShouldSwitchInFodder` to `ShouldSwitch` that checks if the current mon is still useful and a fodder mon exists
```
for (mon in player party)
{
	if (!battlerIsFodder[aiPartyIndex][playerPartyIndex])
		continue;
	else
		fodderId = aiPartyIndex;
}
```
This'd need some extra gates around it obviously to prompt switches that then switch into the wrong mons, but this is the best idea I can come up with.

I frankly don't think this is very practical, as the computation is I think going to be prohibitively expensive to actual use. Open to thoughts or radically different implementation ideas though!

**Comment by AlexOn1ine** [comment]
```
for (mon in AI party)
{
	for (mon in player party)
	{
		for (move in AI mon)
		{
			if (CanWin1v1(aiMon, playerMon, move)
				battlerIsFodder[aiPartyIndex][playerPartyIndex] = FALSE;
			else
				battlerIsFodder[aiPartyIndex][playerPartyIndex] = TRUE;
		}	
	}
}
```
😳

**Comment by Pawkkie** [comment]
Yeah it's obscenely expensive. I don't think there's a way around it if we want to track whether each AI mon is considered fodder at the scale of an entire time. Blerg.

**Comment by Pawkkie** [comment]
Replacing with #6061 to tightly scope the requested feature for future reference
